### DISCUSSION TITLE
/articles/sharepoint-framework-consider-avoiding-declarative-permissions-with-azure-ad-services-in-spfx-projects/

### DISCUSSION THREAD

**Discussion by vtns-comment-importer** [Voitanos/voitanos.io-comments#199] [Comments]
Consider Avoiding Declarative Permissions with Azure AD Services in SharePoint Framework Projects
https://www.andrewconnell.com/articles/sharepoint-framework-consider-avoiding-declarative-permissions-with-azure-ad-services-in-spfx-projects/

**Comment by vtns-comment-importer** [comment]
```header
Originally posted by Sergei Sergeev on 2019-02-13 00:08
```

it's worth mentioning that new isolated web parts feature will solve "tenant" scoped permission issue. Isolated web parts will come in a newer version of SPFx (1.7+) and currently they are in preview.

**Reply by vtns-comment-importer** [reply]
```header
Originally posted by Anton Khritonenkov on 2019-04-16 09:38
```

And still, Isolated web part is an IFrame. It ruins user experience.

**Reply by vtns-comment-importer** [reply]
```header
Originally posted by Andrew Connell on 2019-04-16 10:17
```

You can’t have it both ways… the only way to create a secure client-side component is to isolate it from the hosted domain and to do that, you MUST use an IFRAME (the same way SharePoint Add-ins worked). If you don’t like that you can use the DIV model which includes the component in the hosted page, but you have to trade off isolation.

**Reply by vtns-comment-importer** [reply]
```header
Originally posted by Anton Khritonenkov on 2019-04-16 10:30
```

I know the math. As you say "the same way SharePoint Add-ins worked". That is what we were glad to get rid of when switching to SPFx. Now, as a product developer, I have to face it again. Yes, that is the necessary evil some times, you are right. However, it doesn't change all my concerns about the user experience of those web parts. Hope sometime later, browsers will offer us alternative approaches for isolation with less harm to user experience.

**Reply by vtns-comment-importer** [reply]
```header
Originally posted by Andrew Connell on 2019-04-16 10:45
```

Fair enough... with today's current technology, this is the best we have.

**Reply by juliemturner** [reply]
Just to clarify the issue is not that iFrames ruin the user experience, it's how Microsoft chose to style iFrames in SharePoint pages, etc. If done right it can be a completely fine and responsive experience.

**Comment by vtns-comment-importer** [comment]
```header
Originally posted by Denis Molodtsov on 2020-01-10 13:01
```

Can someone confirm my suspicion, please? Using Azure CLI is a good idea. But, unless I'm mistaken, it requires more permissions. In order to approve API Permissions via the SharePoint Admin Center, it is enough to have a SharePoint Admin role. In order to use Azure CLI to grant consent, we probably need to have a global Azure Admin role. Is that right?

**Reply by vtns-comment-importer** [reply]
```header
Originally posted by Andrew Connell on 2020-01-10 13:19
```

If you are granting the permission via Azure CLI, you need different permissions. The reason is because Azure’s permission model != SharePoint in any way. So… with the Azure CLI, you need the global admin role (for Azure AD). However, you can use the Office 365 CLI and be a SharePoint admin to grant the permission. I’m not entirely sure how that maps the an Azure AD global admin role. The permissions aren’t SharePoint permissions. All you are doing is granting a permission (scope) between two Azure AD apps… SharePoint Admin Center just has a UX for this & a request approve/reject layer on top of Azure AD.

**Reply by vtns-comment-importer** [reply]
```header
Originally posted by Denis Molodtsov on 2020-01-10 13:59
```

Thank you Andrew. My bad: I see you are suggest using Office 365 CLI, not the Azure CLI. I've misread this part. That's actually a great idea. It means it is enough to have a SharePoint admin role to automate the API consent.

**Reply by vtns-comment-importer** [reply]
```header
Originally posted by Andrew Connell on 2020-01-10 14:15
```

Exactly… I’d use the O365 CLI instead of the Azure CLI for this. While I haven’t tried the Azure CLI, I would expect it to work too…

**Comment by vtns-comment-importer** [comment]
```header
Originally posted by Julie Turner on 2020-02-14 09:11
```

I agree that this method would be significantly cleaner and relieve some confusion, however I would strongly recommend that if you're going to go that way as an organization that you have some sort of documentation in a governance plan that this is what your process is (yes I know almost no one does this).

My point is that if someone "else" is trying to determine what permissions have been granted and don't know or think of this work around they're not going to see anything in that list which may inadvertently lead them to thinking that no permissions have been granted, when in fact they have.

I might rather see SharePoint engineering change this from a list of declared items to a "view" of the permission granted in that app registration.

**Comment by ryanshane** [comment]
When i try to run m365 spo serviceprincipal grant list or Get-SPOTenantServicePrincipalPermissionGrants, i receive "Unknown Error" and if i try to run Approve-SPOTenantServicePrincipalPermissionGrant, i get an error that it can not find the service principal? Anyone else experience this? I should be a global administrator. I also tested it on my own tenant and the same happened.

**Reply by andrewconnell** [reply]
You're signing into your SPO tenant's admin center URL, correct? If so, maybe share the whole statement. Hard to help with the limited context provided.

**Reply by ryanshane** [reply]
Hey, sorry, thanks for the reply, i'm very keen to take this approach rather than declarative. I tried the following:

```ps
Connect-SPOService -Url https://{tenant}-admin.sharepoint.com
Approve-SPOTenantServicePrincipalPermissionGrant -Resource "Microsoft Graph" -Scope "User.Read"
```

The connect command was fine, i used a tenant admin account, then authenticator, then it succeeded.
And then when i tried the second line, the error returned is:

```ps
Approve-SPOTenantServicePrincipalPermissionGrant : The service principal for permission request could not be found.
```

Tested with Microsoft.Online.SharePoint.PowerShell version 16.0.24614.12000. I will try the latest version also.

**Reply by ryanshane** [reply]
Hey, despite the error, i can see the request listed on the sharepint admin site and so i will manually approve it here

**Reply by andrewconnell** [reply]
Syntex looks correct, so looks like an error with the SPO POSH module possibly... 

Have you tried using the CLI for Microsoft 365 / [PnP POSH cmdlet](https://pnp.github.io/powershell/cmdlets/Approve-PnPTenantServicePrincipalPermissionRequest.html)?

**Comment by kkazala** [comment]
That’s an interesting one, because according to [Control the SharePoint Online Client service principal
](https://learn.microsoft.com/en-us/sharepoint/dev/spfx/use-aadhttpclient#:~:text=Making%20any%20changes%20directly%20to%20the%20principal%20in%20the%20Microsoft%20Entra%20admin%20center%20can%20result%20in%20unpredictable%20behaviors%20and%20is%20not%20supported), making any changes directly to the SharePoint Online Client Extensibility Web Application Principal
 in the Microsoft Entra admin center can result in unpredictable behaviors and is not supported.
https://github.com/pnp/powershell/discussions/4179

**Reply by andrewconnell** [reply]
You're correct, **but that's not what I'm recommending**. I'm not saying directly update the Entra ID application.

If you use the SharePoint Online PowerShell or CLI for Microsoft 365 commands as I show in this article, you're adding the permission request/grant through the SharePoint Admin Center. That's what I'm recommending.
